gateway-service-impl/src/main/java/org/hypertrace/gateway/service/entity/EntityService.java

```diff
@@ -235,6 +246,61 @@ public BulkUpdateEntitiesResponse bulkUpdateEntities(
     return edsEntityUpdater.bulkUpdateEntities(request, updateExecutionContext);
   }
 
+  private EntitiesRequest addEntityIdsFromInteractionFilter(
+      RequestContext requestContext, EntitiesRequest preProcessedRequest) {
+    Set<EntityKey> entityKeys =
+        this.interactionsFetcher.fetchInteractionsIdsIfNecessary(
+            requestContext, preProcessedRequest);
+    if (!entityKeys.isEmpty()) {
+      Filter entityIdsInFilter =
+          createEntityIdInFilter(requestContext, preProcessedRequest.getEntityType(), entityKeys);
+
+      EntitiesRequest.Builder preProcessRequestBuilder = preProcessedRequest.toBuilder();
+      if (preProcessRequestBuilder.getFilter().equals(Filter.getDefaultInstance())) {
+        preProcessRequestBuilder.setFilter(entityIdsInFilter);
+      } else {
+        preProcessRequestBuilder.setFilter(
+            Filter.newBuilder()
+                .setOperator(Operator.AND)
+                .addChildFilter(preProcessRequestBuilder.getFilter())
+                .addChildFilter(entityIdsInFilter)
+                .build());
+      }
+      preProcessedRequest = preProcessRequestBuilder.build();
+    }
+    return preProcessedRequest;
+  }
+
+  private Filter createEntityIdInFilter(
+      RequestContext requestContext, String entityType, Set<EntityKey> entityKeys) {
+    List<EntityKey> entities = new ArrayList<>(entityKeys);
```

Comment:
These aren't entities, they're keys - please update naming.

Also confused on the next few lines - if we received more keys than we can return entities, why would we drop them? This isn't the only filter - other filters can still bring the result beneath the limit (and even if it was, that's what the limit is for - we shouldn't be enforcing in code)

Reply:
Okay consider this case, lets say there are 5 interaction requests. Now each query will fetch 10k entity Ids. So in the worst case, the entity id filter will have 5 x 10k entity ids as the filter. I put this check to safeguard the 10k limit over the filter

Reply:
Yes, that's a problem - but it's a problem with the general approach that I had mentioned earlier and this isn't a good safeguard for it, since it just breaks the filter logic.

I could just as easily have an edge filter that's super generic that matches everything or almost everything - effectively the edges shouldn't be filtered.

So imagine if I had a querty that said "I want a service with ID = "xyz" and outgoing api calls where the api name matches ".*". By pre-processing the outgoing api calls to get a list of service IDs that match first, and appending that to the filter, if XYZ isn't in the first 10K we'd be forming a filter that returns no results. Hope that makes sense.

Reply:
I have removed this trim method